### PR TITLE
feat(lending): adapter, protocol use async to init config

### PR DESCRIPTION
**Pull request by LuPoYi**
- [x] remove lending protocol token config, use real-time api instead
- [x] make `Adapter` `LendingProtocol` to asynchronous to handle `getReserveTokens`
- [ ] test:unit
- [x] yarn changeset

### CODE REVIEW

File: packages/lending/src/protocols/morphoblue/configs.ts [hunk 1]
```diff
@@ -1,195 +1,30 @@
...
-      },
-    ],
+export const loanTokenPriceFeedMap: Record<number, Record<string, string>> = {
+  [common.ChainId.mainnet]: {
+    [mainnetTokens.WETH.address]: '0x5f4eC3Df9cbd43714FE2740f5E3616155c5b8419',
+    [mainnetTokens.USDC.address]: '0x8fFfFfd4AfB6115b954Bd326cbe7B4BA576818f6',
+    [mainnetTokens.USDT.address]: '0x3E7d1eAB13ad0104d2750B8863b489D65364e32D',
+    [mainnetTokens.PYUSD.address]: '0x8f1dF6D7F2db73eECE86a18b4381F4707b918FB1',
+    [mainnetTokens.DAI.address]: '0xAed0c38402a5d19df6E4c03F4E2DceD6e29c1ee9',
+    [mainnetTokens.USDA.address]: '0x8fFfFfd4AfB6115b954Bd326cbe7B4BA576818f6', // use USDC
```
Comment (LuPoYi) [inline packages/lending/src/protocols/morphoblue/configs.ts:15]: It seems there is no USDA price feed available

File: packages/lending/src/adapter.ts [hunk 2]
```diff
@@ -44,16 +44,26 @@ export class Adapter extends common.Web3Toolkit {
 
   constructor(chainId: number, provider?: providers.Provider) {
     super(chainId, provider);
+  }
+
+  public static async createAdapter(chainId: number, provider?: providers.Provider): Promise<Adapter> {
```
Comment (zodahu) [inline packages/lending/src/adapter.ts:49]: Since this is a breaking change, is it as essential such as `async`? Or can we minimize the impact on integrators when they upgrade the lending-sdk npm?

Comment (LuPoYi) [reply packages/lending/src/adapter.ts:49]: The reason we are doing this is because when we need to create a new adapter, we should initialize multiple protocols, which requires fetching tokens from external sources.

Comment (LuPoYi) [reply packages/lending/src/adapter.ts:49]: Another way is to let integrators call `await adapter.initializeConfig()` themselves
```
    const adapter = new Adapter(chainId, provider);
    await adapter.initializeConfig();
    await adapter.getPortfolio(user.address, protocolId, marketId);
```

File: packages/lending/src/protocols/aave-v2/configs.ts [hunk 3]
```diff
@@ -109,128 +34,30 @@ export const configs: Config[] = [
...
 export function getContractAddress(chainId: number, name: ContractName) {
-  return configMap[chainId].contractMap[name];
+  const { contractMap } = configs.find((configs) => configs.chainId === chainId)!;
+  return contractMap[name];
 }
 
-export function toAToken(chainId: number, token: common.Token) {
-  return reserveMap[chainId][token.wrapped.address].aToken;
-}
+const depositDisableMap: Record<number, string[]> = {
```
Comment (zodahu) [inline packages/lending/src/protocols/aave-v2/configs.ts]:
```suggestion
const supplyDisabledMap: Record<number, string[]> = {
```

Can we use `supply` instead of `deposit` as we almost use `supply` term in lending-sdk for portfolio and zap-supply naming etc.

File: packages/lending/src/protocols/aave-v2/configs.ts [hunk 4]
```diff
@@ -109,128 +34,30 @@ export const configs: Config[] = [
...
+  [common.ChainId.mainnet]: [],
+  [common.ChainId.polygon]: [],
+};
 
-export function isAToken(chainId: number, token: common.Token): boolean {
-  const aToken = reserveMap[chainId][token.address].aToken;
-  return aToken.is(token);
-}
+const borrowDisableMap: Record<number, string[]> = {
```
Comment (zodahu) [inline packages/lending/src/protocols/aave-v2/configs.ts]:
```suggestion
const borrowDisabledMap: Record<number, string[]> = {
```

File: packages/lending/src/protocols/aave-v2/configs.ts [hunk 5]
```diff
@@ -109,128 +34,30 @@ export const configs: Config[] = [
...
+  [common.ChainId.mainnet]: [common.mainnetTokens.AAVE.address, common.mainnetTokens.stETH.address],
+  [common.ChainId.polygon]: [common.polygonTokens.AAVE.address],
+};
 
-export function toToken(chainId: number, aToken: common.Token, unwrap = true) {
-  const asset = reserveMap[chainId][aToken.address].asset;
-  return unwrap ? asset.unwrapped : asset;
-}
+export const isTokenForDeposit = (chainId: number, token: common.Token) => {
```
Comment (zodahu) [inline packages/lending/src/protocols/aave-v2/configs.ts]:
```suggestion
export const isSupplyEnabled = (chainId: number, token: common.Token) => {
```

File: packages/lending/src/protocols/aave-v2/configs.ts [hunk 6]
```diff
@@ -109,128 +34,30 @@ export const configs: Config[] = [
...
+  return !depositDisableMap[chainId].includes(token.address);
+};
+
+export const isTokenForBorrow = (chainId: number, token: common.Token) => {
```
Comment (zodahu) [inline packages/lending/src/protocols/aave-v2/configs.ts]:
```suggestion
export const isBorrowEnabled = (chainId: number, token: common.Token) => {
```

File: packages/lending/src/protocols/aave-v2/tokens/data/mainnet.json [hunk 7]
```diff
@@ -1,74 +1,32 @@
...
-    "symbol": "USDT",
-    "name": "Tether USD"
-  },
   "aUSDT": {
     "chainId": 1,
     "address": "0x3Ed3B47Dd13EC9a98b44e6204A523E766B225811",
     "decimals": 6,
     "symbol": "aUSDT",
     "name": "Aave interest bearing USDT"
   },
```
Comment (zodahu) [inline packages/lending/src/protocols/aave-v2/tokens/data/mainnet.json]: Is it possible we remove all the unused testing tokens for in all `/data/*.json`? Like what we did in `protocolink-logics` repo.

File: packages/lending/src/adapter.ts (hunk 2 above)
Comment (zodahu) [reply packages/lending/src/adapter.ts:49]: Do you think we should increase the major version number?

File: packages/lending/src/protocols/aave-v2/configs.ts [hunk 8]
```diff
@@ -109,128 +34,30 @@ export const configs: Config[] = [
...
+};
 
-export function isAToken(chainId: number, token: common.Token): boolean {
-  const aToken = reserveMap[chainId][token.address].aToken;
-  return aToken.is(token);
-}
+const borrowDisableMap: Record<number, string[]> = {
+  [common.ChainId.mainnet]: [common.mainnetTokens.AAVE.address, common.mainnetTokens.stETH.address],
+  [common.ChainId.polygon]: [common.polygonTokens.AAVE.address],
+};
```
Comment (zodahu) [inline packages/lending/src/protocols/aave-v2/configs.ts]:
```suggestion
  [common.ChainId.mainnet]: [common.mainnetTokens.stETH.address],
  [common.ChainId.polygon]: [],
  [common.ChainId.avalanche]: [],
};
```

`AAVE` is not in `service.getReserveTokens()`

File: packages/lending/src/protocols/aave-v2/configs.ts [hunk 9]
```diff
@@ -109,128 +34,30 @@ export const configs: Config[] = [
...
+  const { contractMap } = configs.find((configs) => configs.chainId === chainId)!;
+  return contractMap[name];
 }
 
-export function toAToken(chainId: number, token: common.Token) {
-  return reserveMap[chainId][token.wrapped.address].aToken;
-}
+const depositDisableMap: Record<number, string[]> = {
+  [common.ChainId.mainnet]: [],
+  [common.ChainId.polygon]: [],
```
Comment (zodahu) [inline packages/lending/src/protocols/aave-v2/configs.ts:58]:
```suggestion
  [common.ChainId.mainnet]: [],
  [common.ChainId.polygon]: [],
  [common.ChainId.avalanche]: [],
```

File: packages/lending/src/adapter.ts (hunk 2 above)
Comment (zodahu) [reply packages/lending/src/adapter.ts:49]: I believe we need major number change as we also remove functions such as `exported config.reserves`, original `getMarketName` now is `getProtocolName` etc.

File: packages/lending/src/protocols/radiant-v2/lending-protocol.ts [hunk 10]
```diff
@@ -101,14 +126,44 @@ export class LendingProtocol extends Protocol {
...
 
-  getMarketName() {
+  getProtocolName() {
     return DISPLAY_NAME;
   }
 
+  getMarketName() {
+    return this.market.id;
+  }
+
```
Comment (zodahu) [inline packages/lending/src/protocols/radiant-v2/lending-protocol.ts]: Why we introduce this change on abstract `protocol.ts`? And it seems it should be `getMarkeId`?

File: packages/lending/src/protocols/aave-v3/configs.ts [hunk 11]
```diff
@@ -634,123 +89,42 @@ export const configs: Config[] = [
...
+const depositDisableMap: Record<number, string[]> = {
+  [common.ChainId.mainnet]: [mainnetTokens.GHO.address],
+  [common.ChainId.gnosis]: [],
+  [common.ChainId.polygon]: [],
+  [common.ChainId.base]: [],
+  [common.ChainId.arbitrum]: [],
+  [common.ChainId.optimism]: [],
+  [common.ChainId.avalanche]: [],
+  [common.ChainId.metis]: [],
+};
```
Comment (zodahu) [inline packages/lending/src/protocols/aave-v3/configs.ts]: Sort supply and borrow maps

File: packages/lending/src/protocols/aave-v3/configs.ts [hunk 12]
```diff
@@ -634,123 +89,42 @@ export const configs: Config[] = [
...
 
-export function toToken(chainId: number, aToken: common.Token, unwrap = true) {
-  const asset = reserveMap[chainId][aToken.address].asset;
-  return unwrap ? asset.unwrapped : asset;
-}
+const borrowDisableMap: Record<number, string[]> = {
+  [common.ChainId.mainnet]: [mainnetTokens.AAVE.address, mainnetTokens.STG.address, mainnetTokens.FXS.address],
```
Comment (zodahu) [inline packages/lending/src/protocols/aave-v3/configs.ts]:
```suggestion
  [common.ChainId.mainnet]: [mainnetTokens.AAVE.address],
```
`STG` and `FXS` are not in `getReserves` list, i.e., they are frozen on Aave UI.

File: packages/lending/src/protocols/aave-v3/lending-protocol.test.ts [hunk 13]
```diff
@@ -917,2607 +917,20 @@ describe('Test Aave V3 LendingProtocol', function () {
           ],
         },
       },
-      {
-        chainId: common.ChainId.gnosis,
-        account: '0xd0214b7e8a7821a5cb07024bc00d64ece8cc1067',
-        blockTag: 32001300,
-        expected: {
```
Comment (zodahu) [inline packages/lending/src/protocols/aave-v3/lending-protocol.test.ts]: Why we delete all testing on all chains except mainnet?

File: packages/lending/src/protocols/compound-v3/lending-protocol.test.ts [hunk 14]
```diff
@@ -522,6 +522,12 @@ describe('Test Compound V3 LendingProtocol', function () {
           ],
         },
       },
+      {
+        chainId: common.ChainId.arbitrum,
+        marketId: logics.compoundv3.MarketId.USDCe,
+        account: '0xA62315902fAADC69F898cc8B85F86FfD1F6aAeD8',
+        expected: {},
+      },
```
Comment (zodahu) [inline packages/lending/src/protocols/compound-v3/lending-protocol.test.ts]: Why was this added?

File: packages/lending/src/protocols/aave-v2/tokens/data/mainnet.json (hunk 7 above)
Comment (zodahu) [reply packages/lending/src/protocols/aave-v2/tokens/data/mainnet.json]: Or is it possible to use `protocolink-logics`' `/data/*.json` so that we can keep no `/data/*.json` for each protocol in lending-sdk?

File: packages/lending/src/protocols/radiant-v2/configs.ts [hunk 15]
```diff
@@ -31,235 +23,46 @@ export const configs: Config[] = [
...
+};
 
-export function isRToken(chainId: number, token: common.Token): boolean {
-  const rToken = reserveMap[chainId][token.address].rToken;
-  return rToken.is(token);
-}
+const borrowDisableMap: Record<number, string[]> = {
+  [common.ChainId.mainnet]: [],
+  [common.ChainId.bnb]: [],
+  [common.ChainId.arbitrum]: [],
```
Comment (zodahu) [inline packages/lending/src/protocols/radiant-v2/configs.ts]: Arbitrum had onboard gmETH and gmBTC as collateral but cannot borrow them on 6/14.

https://dao.radiant.capital/#/proposal/0x8216d4b089c0e787ef383683b17ce2af026caa2bba986ddbddfa178094bf854b

File: packages/lending/src/protocols/spark/configs.ts [hunk 16]
```diff
@@ -89,123 +35,30 @@ export const configs: Config[] = [
...
+  [common.ChainId.gnosis]: [],
+};
 
-export function toToken(chainId: number, aToken: common.Token, unwrap = true) {
-  const asset = reserveMap[chainId][aToken.address].asset;
-  return unwrap ? asset.unwrapped : asset;
-}
+const borrowDisableMap: Record<number, string[]> = {
+  [common.ChainId.mainnet]: [mainnetTokens.sDAI.address],
+  [common.ChainId.gnosis]: [gnosisTokens.GNO.address],
```
Comment (zodahu) [inline packages/lending/src/protocols/spark/configs.ts]:
```suggestion
  [common.ChainId.gnosis]: [gnosisTokens.GNO.address, gnosisTokens.sDAI.address],
```
image

File: packages/lending/src/adapter.ts (hunk 2 above)
Comment (LuPoYi) [reply packages/lending/src/adapter.ts:49]: updated changeset file to `major`, so that will `2.0.0`

File: packages/lending/src/protocols/aave-v2/configs.ts (hunk 3 above)
Comment (LuPoYi) [reply packages/lending/src/protocols/aave-v2/configs.ts]: fixed, let's keep it the same term

File: packages/lending/src/protocols/aave-v2/configs.ts (hunk 4 above)
Comment (LuPoYi) [reply packages/lending/src/protocols/aave-v2/configs.ts]: fixed

File: packages/lending/src/protocols/aave-v2/configs.ts (hunk 5 above)
Comment (LuPoYi) [reply packages/lending/src/protocols/aave-v2/configs.ts]: fixed

File: packages/lending/src/protocols/aave-v2/configs.ts (hunk 6 above)
Comment (LuPoYi) [reply packages/lending/src/protocols/aave-v2/configs.ts]: fixed

File: packages/lending/src/protocols/aave-v2/configs.ts (hunk 8 above)
Comment (LuPoYi) [reply packages/lending/src/protocols/aave-v2/configs.ts]: removed

File: packages/lending/src/protocols/aave-v2/configs.ts (hunk 9 above)
Comment (LuPoYi) [reply packages/lending/src/protocols/aave-v2/configs.ts:58]: added, and also updated config part

File: packages/lending/src/protocols/radiant-v2/lending-protocol.ts (hunk 10 above)
Comment (LuPoYi) [reply packages/lending/src/protocols/radiant-v2/lending-protocol.ts]: There are three different protocols like this. We need to get the protocol name and market name to display
`getMarketName` for single market protocol like radiantv2 is not very useful

```
Protocol id: radiant-v2
Protocol name: Radiant V2
makret name: mainnet

Protocol id: compound-v3
Protocol name: Compound V3
makret name: USDC

Protocol id: morphoblue
Protocol name: Morpho Blue
makret name: wstETH/WETH
```

File: packages/lending/src/protocols/aave-v3/configs.ts (hunk 11 above)
Comment (LuPoYi) [reply packages/lending/src/protocols/aave-v3/configs.ts]: sorted

File: packages/lending/src/protocols/aave-v3/configs.ts (hunk 12 above)
Comment (LuPoYi) [reply packages/lending/src/protocols/aave-v3/configs.ts]: removed

File: packages/lending/src/protocols/aave-v3/lending-protocol.test.ts (hunk 13 above)
Comment (LuPoYi) [reply packages/lending/src/protocols/aave-v3/lending-protocol.test.ts]: Accidentally removed for testing, it has been added back

File: packages/lending/src/protocols/compound-v3/lending-protocol.test.ts (hunk 14 above)
Comment (LuPoYi) [reply packages/lending/src/protocols/compound-v3/lending-protocol.test.ts]: added for testing and forgotten to be removed. It has now been removed

File: packages/lending/src/protocols/radiant-v2/configs.ts (hunk 15 above)
Comment (LuPoYi) [reply packages/lending/src/protocols/radiant-v2/configs.ts]: added gmETH, gmBTC to borrowDisabledMap

File: packages/lending/src/protocols/spark/configs.ts (hunk 16 above)
Comment (LuPoYi) [reply packages/lending/src/protocols/spark/configs.ts]: fixed

File: packages/lending/src/protocols/aave-v2/tokens/data/mainnet.json (hunk 7 above)
Comment (LuPoYi) [reply packages/lending/src/protocols/aave-v2/tokens/data/mainnet.json]: This seems feasible, only radiant's gmETH and gmBTC would need to be set, the rest wouldn't need tokens.json.

Comment (LuPoYi) [reply packages/lending/src/protocols/aave-v2/tokens/data/mainnet.json]: since in `protocolink-logics`, we already do something like this
```
// protocolink-logics
export const arbitrumTokens = {
  ...common.toTokenMap<ArbitrumTokenSymbols>(arbitrumTokensJSON),
  ...common.arbitrumTokens,
};
```

I think we can remove `*.json` and keep `src/protocols/radiant-v2/tokens/index.ts`, and change to
```
// lending sdk
export const arbitrumTokens = logics.radiantv2.arbitrumTokens;

// if we still need other token
export const arbitrumTokens = {...logics.radiantv2.arbitrumTokens, ...otherTokens};
```

File: packages/lending/src/protocols/radiant-v2/configs.ts (hunk 15 above)
Comment (zodahu) [reply packages/lending/src/protocols/radiant-v2/configs.ts]: gmBTC address is `0x47c031236e19d024b42f8AE6780E44A573170703`.

File: packages/lending/src/protocols/aave-v2/tokens/data/mainnet.json (hunk 7 above)
Comment (LuPoYi) [reply packages/lending/src/protocols/aave-v2/tokens/data/mainnet.json]: https://github.com/dinngo/protocolink-js-sdk/pull/63/commits/34c81a23beeb102d756da8e44cfa5f3be376b85f

If `protocolink-logics` has defined tokens, then use `logics.xxxxxx.yyyyyyTokens`, otherwise, use the `common.yyyyyTokens`

File: packages/lending/src/protocols/aave-v2/adapter.test.ts [hunk 17]
```diff
@@ -266,6 +271,7 @@ describe('Test Adapter for Aave V2', function () {
       expect(Number(destAmount)).to.be.greaterThan(0);
       expect(afterPortfolio.totalSupplyUSD).to.be.eq(0);
 
+      console.log(logics);
```
Comment (zodahu) [inline packages/lending/src/protocols/aave-v2/adapter.test.ts]: Remove debug log

File: packages/lending/src/protocols/aave-v3/lending-protocol.test.ts [hunk 18]
```diff
@@ -3475,49 +3840,35 @@ describe('Test Aave V3 LendingProtocol', function () {
...
             },
           ],
         },
       },
     ];
 
     testCases.forEach(({ chainId, account, blockTag, expected }) => {
-      it(`${common.toNetworkId(chainId)} market`, async function () {
-        const protocol = new LendingProtocol(chainId);
+      it(`${common.toNetworkId(chainId)} market ${blockTag}`, async function () {
```
Comment (zodahu) [inline packages/lending/src/protocols/aave-v3/lending-protocol.test.ts]:
```suggestion
      it(`${common.toNetworkId(chainId)} market`, async function () {
```
Is debug log?

**Comment by zodahu** [comment]
LGTM
